### ISSUE TITLE
Debian 7 Makecheck Fails - (src/core/sock.c:830) Endpoint is stopped. Now we can safely deallocate it. 

### ISSUE THREAD

**Issue by marchon** [nanomsg/nanomsg#368]
Confirmed on a fresh Debian 7 64 bit Install 

Linux version 3.18.3-x86_64-linode51 (maker@build.linode.com) (gcc version 4.4.5 (Debian 4.4.5-8) ) #1 SMP Fri Jan 23 09:57:22 EST 2015

#Error 
## in commit 1ae64bb578a74bd0912698bb4771cbc6e2e69d07 titled "Better Error Reporting"  
https://github.com/nanomsg/nanomsg/commit/1ae64bb578a74bd0912698bb4771cbc6e2e69d07

sustrik added debugging to dump the src and type variables to compare them against nanomsg constants
```c 
#define NN_SOCK_SRC_EP 1                              //src/core/sock.c: 
#define NN_EP_STOPPED 1                                //src/core/ep.h: 
#define NN_SOCK_STATE_STOPPING_EPS 4   // src/core/sock.c:
```

I ran make check on a fresh Debian 7 install - and here is the error output

``` bash
root@li900-150:~/nanomsg# make check > report.txt
src=0 type=33988
Assertion failed: src == NN_SOCK_SRC_EP && type == NN_EP_STOPPED
(src/core/sock.c:830)
```

the code in question is reporting incorrect values for both src and type

both of which are passed into the function 

```
static void nn_sock_shutdown(struct nn_fsm *self, int src, int type, void *srcptr)  
```
and are not changed before this comparison. 

src=0 // should be 1
type=33988 // should be 1 

``` c
   if (nn_slow (sock->state == NN_SOCK_STATE_STOPPING_EPS)) {
        /*  Endpoint is stopped. Now we can safely deallocate it. */

        if (!(src == NN_SOCK_SRC_EP && type == NN_EP_STOPPED)) {
            fprintf (stderr, "src=%d type=%d\n", (int) src, (int) type);
            nn_assert (src == NN_SOCK_SRC_EP && type == NN_EP_STOPPED);
        }

        ep = (struct nn_ep*) srcptr;
        nn_list_erase (&sock->sdeps, &ep->item);
        nn_ep_term (ep);
        nn_free (ep);
```

Machine Configuration 

```bash 
root@li900-151:~/nanomsg# uname -mrs                                                                    
Linux 3.18.3-x86_64-linode51 x86_64                                                                     
root@li900-151:~/nanomsg# uname -a                                                                      
Linux li900-151 3.18.3-x86_64-linode51 #1 SMP Fri Jan 23 09:57:22 EST 2015 x86_64 GNU/Linux             
root@li900-151:~/nanomsg# cat /proc/version                                                             
Linux version 3.18.3-x86_64-linode51 (maker@build.linode.com) (gcc version 4.4.5 (Debian 4.4.5-8) ) #1 S
MP Fri Jan 23 09:57:22 EST 2015                                                                         
root@li900-151:~/nanomsg# 
```

**Comment by marchon** [comment]
Did not fail on a Test VirtualBox with older kernel.

```
===================
All 33 tests passed
===================
make[1]: Leaving directory `/home/vagrant/nanomsg'
root@vagrant-debian-wheezy:/home/vagrant/nanomsg#  uname -mrs        
Linux 3.2.0-4-amd64 x86_64
root@vagrant-debian-wheezy:/home/vagrant/nanomsg# uname -a    
Linux vagrant-debian-wheezy 3.2.0-4-amd64 #1 SMP Debian 3.2.63-2+deb7u2 x86_64 GNU/Linux
root@vagrant-debian-wheezy:/home/vagrant/nanomsg# cat /proc/version
Linux version 3.2.0-4-amd64 (debian-kernel@lists.debian.org) (gcc version 4.6.3 (Debian 4.6.3-14) ) #1 SMP Debian 3.2.63-2+deb7u2
```

**Comment by marchon** [comment]
It is not the kernel version, as I have upgraded the kernel from Linux 3.2.0-4-amd64 x86_64 to Linux 3.18.0-10-exton x86_64



```
PASS: tests/inproc
PASS: tests/inproc_shutdown
PASS: tests/ipc
PASS: tests/ipc_shutdown
PASS: tests/ipc_stress
PASS: tests/tcp
PASS: tests/tcp_shutdown
PASS: tests/ws
PASS: tests/tcpmux
PASS: tests/pair
PASS: tests/pubsub
PASS: tests/reqrep
PASS: tests/pipeline
PASS: tests/survey
PASS: tests/bus
PASS: tests/block
PASS: tests/term
PASS: tests/timeo
PASS: tests/iovec
PASS: tests/msg
PASS: tests/prio
PASS: tests/poll
PASS: tests/device
PASS: tests/emfile
PASS: tests/domain
PASS: tests/trie
PASS: tests/list
PASS: tests/hash
PASS: tests/symbol
PASS: tests/separation
PASS: tests/zerocopy
PASS: tests/shutdown
PASS: tests/cmsg
make[3]: Entering directory '/home/vagrant/nanomsg'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/home/vagrant/nanomsg'
============================================================================
Testsuite summary for nanomsg 0.5-beta-73-g2da0726
============================================================================
# TOTAL: 33
# PASS:  33
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
make[2]: Leaving directory '/home/vagrant/nanomsg'
make[1]: Leaving directory '/home/vagrant/nanomsg'
root@vagrant-debian-wheezy:/home/vagrant/nanomsg# uname -mrs
Linux 3.18.0-10-exton x86_64
root@vagrant-debian-wheezy:/home/vagrant/nanomsg# uname -a
Linux vagrant-debian-wheezy 3.18.0-10-exton #11 SMP Fri Jan 23 17:09:14 CET 2015 x86_64 GNU/Linux
root@vagrant-debian-wheezy:/home/vagrant/nanomsg# cat /proc/version
Linux version 3.18.0-10-exton (root@mate-u) (gcc version 4.9.1 (Ubuntu 4.9.1-16ubuntu6) ) #11 SMP Fri Jan 23 17:09:14 CET 2015
```

**Comment by marchon** [comment]
It is also not an IPV6 Issues

**Comment by marchon** [comment]
I dropped the Kernel Version on the initial failing machine to Linux 2.6.39.1-x86_64 and the problem persists.  I added debugging code to log the src and type value of every call to

**Comment by wirebirdlabs** [comment]
Seems related to #269 

My current workaround is to ignore messages of type NN_PIPE_IN and NN_PIPE_OUT while a socket is in its final stages of shutting down -- something like this: https://github.com/wirebirdlabs/featherweight-nanomsg/blob/8516fc4285ffb3dfcf5d9e9319f1490b29d800ca/src/core/sock.c#L870-L879

**Comment by wirebirdlabs** [comment]
> ignore messages of type NN_PIPE_IN and NN_PIPE_OUT

Related, @TTimo does similarly: https://github.com/TTimo/nanomsg/commit/518d3ebd02e0ce1a7cc77f8611afeb6c23f8bfe6#commitcomment-9266560

**Comment by gdamore** [comment]
This was fixed in 

commit 7555f49e83ff414133f77e46389f8e5632eeea1f
Author: Garrett D'Amore <garrett@damore.org>
Date:   Fri Oct 23 08:01:22 2015 -0700

    fixes #469 assertion failure in ipc_shutdown (really)
